decision-records/008-reset-the-javascript-api.md

```diff
@@ -0,0 +1,72 @@
+# Reset the JavaScript API for our components
+
+## Decision
+
+We will mark all class and instance properties other than the constructor and the init method as deprecated as part of a v4.x release.
+
+In v5.0 we will then add JSDoc annotations to mark the deprecated properties as private.
+
+In future releases we will then gradually introduce a new public JavaScript API for components, which is:
+
+- based on the needs of users
+- clearly defined and documented
+- comprehensively tested
+
+**Date:** 2023-07-20
+
+**Status:** Accepted
+
+## Rationale
+
+The JavaScript API for our components has not been designed to be called externally and does not clearly differentiate between public and private methods.
+
+Many of the methods exposed on the classes or objects can only be called internally as they need to be passed internal state, or are designed to be used internally by the component as event handlers.
+
+We also have little or no test coverage for calling these methods externally.
+
+This means that:
+
+- it's unclear for users whether they are allowed to call or interact with each property
+- it's difficult for us to judge what constitutes a breaking change
+
+'Resetting' the public API by marking everything else as private gives us the freedom to refactor the internals of the component before then introducing a properly designed, tested and documented public API.
+
+## Risks and constraints
+
+Users who are relying on being able to access now private class or instance properties from from their own code will need to either:
+
+- wait until we introduce an equivalent way to achieve the same thing using the public API before they can upgrade beyond v5
+- continue to use the private properties 'at risk', and manually test or verify changes between releases
+
+## Alternatives considered
```

Comment:
Do we need to explore things like this?

1. Using [actually-private class fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) with a Babel transform
2. Using [JSDoc `@protected`](https://jsdoc.app/tags-protected.html) for properties an extended class could access, but still externally private

Reply:
I don't think so, only because they're not mutually exclusive in the long run – we're choosing to use `@private` annotations now, but we could still choose to make things actually-private or add `@protected` to the mix in the future without needing to revisit this decision.

Reply:
Yeah I do agree

But is it worth documenting that we'll knowingly mark public class fields as `@private`?

E.g. JSDoc compiler [warning about `new Component().i18n`](https://github.com/alphagov/govuk-frontend/blob/d9f593322e1722a6d6b80443f9829ad4e38bfe4c/packages/govuk-frontend-review/typedoc.config.js#L19)

Reply:
I'm not sure of the value of marking individual properties as internal – public or protected seem appropriate there.

Internal is useful for the I18n class because it has public properties (because they need to be called from outside of the class) but we want to differentiate it from classes designed to be called outside of the package.

Reply:
Do `private` or `public` actually make sense on functions outside of the context of a class? If so, what do those annotations mean?

The way I'm thinking about it in my head:

- `public` and `private` relate to access in the context of the encapsulation of a class (whether the method should be called from outside of the class)
- `internal` relates to access in the context of the package (so we can call the public bits of the internal `I18n` class from elsewhere in GOV.UK Frontend, but we don't want users to)

Reply:
Yeah spot on

You've only said "**'internal' classes**" but it should be "**'internal' classes, helpers and utilities**" etc?